File: benchmarks/tasks/pipeline/pipeline.go
// Message passing: four producer goroutines each send fifty thousand numbers
// into one channel, and main receives all of them and adds them up.

package main

import "fmt"

const producers = 4
const each = 50000

func main() {
	ch := make(chan int64, 1024)
	for p := int64(0); p < producers; p++ {
		go func(p int64) {
			for i := int64(0); i < each; i++ {
				ch <- p*each + i
			}
		}(p)
	}
	var total int64
	for n := 0; n < producers*each; n++ {
		total += <-ch
	}
	fmt.Println(total)
}

File: benchmarks/tasks/pipeline/pipeline.rs
// Message passing: four producer threads each send fifty thousand numbers into
// one channel, and the main thread receives all of them and adds them up.

use std::sync::mpsc;
use std::thread;

const PRODUCERS: i64 = 4;
const EACH: i64 = 50_000;

fn main() {
    let (tx, rx) = mpsc::channel::<i64>();
    for p in 0..PRODUCERS {
        let tx = tx.clone();
        thread::spawn(move || {
            for i in 0..EACH {
                tx.send(p * EACH + i).unwrap();
            }
        });
    }
    drop(tx);
    let total: i64 = rx.iter().sum();
    println!("{total}");
}
